src/compile/compile.ts

```diff
@@ -67,6 +65,9 @@ function getTopLevelProperties(topLevelSpec: TopLevel<any>, config: Config) {
 function assembleTopLevelModel(model: Model, topLevelProperties: TopLevelProperties) {
   // TODO: change type to become VgSpec
 
+  // optimize the datafow once
+  optimizeDataflow(model.component.data);
```

Comment:
(Not that this method returns value, so looking at it in `compile.ts` might still feel like `assembleTopLevelModel` shouldn't cause side-effect.